Setup.bat

```diff
@@ -27,14 +36,58 @@ call :MarkStartOfBlock "Setup the git hooks"
 call :MarkEndOfBlock "Setup the git hooks"
 
 call :MarkStartOfBlock "Check dependencies"
-    if not defined UNREAL_HOME (
-        echo Error: Please set UNREAL_HOME environment variable to point to the Unreal Engine folder.
+
+    rem Find the Unreal Engine used to build this project.
+    set UNREAL_ENGINE=""
+
+    rem Get the Unreal Engine used by this project by querying the registry for the engine association found in the .uproject.
+    for /f "delims=" %%A in (' powershell -Command "Get-Childitem -Path %ProjectDirectory% -Recurse -Include *.uproject -File | %% {$_.FullName}" ') do set UPROJECT="%%A"
+    
+    rem If the regex failed then it will return a string containing only a space.
+    if %UPROJECT%==" " (
+        echo Error: Could not find uproject. Please make sure you have passed in the project directory correctly.
         pause
         exit /b 1
     )
 
+    echo Using uproject: %UPROJECT%
+
+    rem Get the Engine Association from the uproject.
+    for /f "delims=" %%A in (' powershell -Command "Select-String -Pattern '..{8}-.{4}-.{4}-.{4}-.{12}.' %UPROJECT% -AllMatches | %% { $_.Matches } | %% { $_.Value }" ') do set ENGINE_ASSOCIATION="%%A"
+    
+    echo Engine association for uproject is: %ENGINE_ASSOCIATION%
+
+    if not %ENGINE_ASSOCIATION%==" " (
+        rem Query the registry for the path to the Unreal Engine using the engine associtation.
+        for /f "usebackq tokens=3*" %%A in (`reg query "HKCU\Software\Epic Games\Unreal Engine\Builds" /v %ENGINE_ASSOCIATION%`) do (
+            set UNREAL_ENGINE="%%A"
+        )
+    )
+
+    rem If there was no engine association then we need to climb the directory path to find the Engine.
+    if %UNREAL_ENGINE%=="" (
+        :climb_parent_directory
+        cd ..
+        if exist Engine (

```

Comment:
What if there is another file or directory called `Engine`?

Reply:
So if we're actually looking for an Engine at this point (no project association set in the project file) then the users game MUST be placed inside of the UnrealEngine. If they have their game project nested in a folder which contains another Engine folder it's a very messy setup. To make it more robust though I've added an additional check for `Engine\Build\Build.version` which is included in all Unreal Engine checkouts. It's never going to be perfect but it will help in 99% of cases I hope.